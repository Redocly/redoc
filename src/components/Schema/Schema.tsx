import { observer } from 'mobx-react';
import * as React from 'react';

import { RecursiveLabel, TypeName, TypeTitle } from '../../common-elements/fields';
import { FieldDetails } from '../Fields/FieldDetails';

import { FieldModel, SchemaModel } from '../../services/models';

import { ArraySchema } from './ArraySchema';
import { ObjectSchema } from './ObjectSchema';
import { OneOfSchema } from './OneOfSchema';

import { l } from '../../services/Labels';

export interface SchemaOptions {
  showTitle?: boolean;
  skipReadOnly?: boolean;
  skipWriteOnly?: boolean;
}

export interface SchemaProps extends SchemaOptions {
  schema: SchemaModel;
}

@observer
export class Schema extends React.Component<Partial<SchemaProps>> {
  render() {
    const { schema } = this.props;
    if (!schema) {
      return <em> Schema not provided </em>;
    }
    const { type, oneOf, discriminatorProp, isCircular } = schema;

    if (isCircular) {
      return (
        <div>
          <TypeName>{schema.displayType}</TypeName>
          {schema.title && <TypeTitle> {schema.title} </TypeTitle>}
          <RecursiveLabel> {l('recursive')} </RecursiveLabel>
        </div>
      );
    }

    if (discriminatorProp !== undefined) {
      if (!oneOf || !oneOf.length) {
        console.warn(
          `Looks like you are using discriminator wrong: you don't have any definition inherited from the ${schema.title}`,
        );
        return null;
      }
      return (
        <ObjectSchema
          {...{ ...this.props, schema: oneOf![schema.activeOneOf] }}
          discriminator={{
            fieldName: discriminatorProp,
            parentSchema: schema,
          }}
        />
      );
    }

    if (oneOf !== undefined) {
      return <OneOfSchema schema={schema} {...this.props} />;
    }

    const types = Array.isArray(type) ? type : [type];
    if (types.includes('object')) {
      if (schema.fields?.length) {
        return <ObjectSchema {...(this.props as any)} />;
      }
    } else if (types.includes('array')) {
      return <ArraySchema {...(this.props as any)} />;
    }

    // TODO: maybe adjust FieldDetails to accept schema
    const field = ({
      schema,
      name: '',
      required: false,
      description: schema.description,
      externalDocs: schema.externalDocs,
      deprecated: false,
      toggle: () => null,
      expanded: false,
    } as any) as FieldModel; // cast needed for hot-loader to not fail

    return (
      <div>
        <FieldDetails field={field} />
      </div>
    );
  }
}
