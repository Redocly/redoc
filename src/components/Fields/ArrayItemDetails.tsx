import * as React from 'react';
import { TypeFormat, TypePrefix } from '../../common-elements/fields';
import { ConstraintsView } from './FieldContstraints';
import { Pattern } from './Pattern';
import { SchemaModel } from '../../services';
import styled from '../../styled-components';

export function ArrayItemDetails({ schema }: { schema: SchemaModel }) {
  if (!schema || (schema.type === 'string' && !schema.constraints.length)) return null;

  return (
    <Wrapper>
      [ items
      {schema.displayFormat && <TypeFormat>{` &lt;${schema.displayFormat}&gt; `}</TypeFormat>}
      <ConstraintsView constraints={schema.constraints} />
      <Pattern schema={schema} />
      {schema.items && <ArrayItemDetails schema={schema.items} />} ]
    </Wrapper>
  );
}

const Wrapper = styled(TypePrefix)`
  margin: 0 5px;
  vertical-align: text-top;
`;
