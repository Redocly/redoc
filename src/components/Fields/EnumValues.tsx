import * as React from 'react';
import { ExampleValue, FieldLabel } from '../../common-elements/fields';

import { l } from '../../services/Labels';
import { OptionsContext } from '../OptionsProvider';
import styled from '../../styled-components';
import { RedocRawOptions } from '../../services/RedocNormalizedOptions';

export interface EnumValuesProps {
  values: string[];
  type: string | string[];
}

export interface EnumValuesState {
  collapsed: boolean;
}

export class EnumValues extends React.PureComponent<EnumValuesProps, EnumValuesState> {
  state: EnumValuesState = {
    collapsed: true,
  };

  static contextType = OptionsContext;

  private toggle() {
    this.setState({ collapsed: !this.state.collapsed });
  }

  render() {
    const { values, type } = this.props;
    const { collapsed } = this.state;

    // TODO: provide context interface in more elegant way
    const { enumSkipQuotes, maxDisplayedEnumValues } = this.context as RedocRawOptions;

    if (!values.length) {
      return null;
    }

    const displayedItems =
      this.state.collapsed && maxDisplayedEnumValues
        ? values.slice(0, maxDisplayedEnumValues)
        : values;

    const showToggleButton = maxDisplayedEnumValues
      ? values.length > maxDisplayedEnumValues
      : false;

    const toggleButtonText = maxDisplayedEnumValues
      ? collapsed
        ? `â€¦ ${values.length - maxDisplayedEnumValues} more`
        : 'Hide'
      : '';

    return (
      <div>
        <FieldLabel>
          {type === 'array' ? l('enumArray') : ''}{' '}
          {values.length === 1 ? l('enumSingleValue') : l('enum')}:
        </FieldLabel>{' '}
        {displayedItems.map((value, idx) => {
          const exampleValue = enumSkipQuotes ? String(value) : JSON.stringify(value);
          return (
            <React.Fragment key={idx}>
              <ExampleValue>{exampleValue}</ExampleValue>{' '}
            </React.Fragment>
          );
        })}
        {showToggleButton ? (
          <ToggleButton
            onClick={() => {
              this.toggle();
            }}
          >
            {toggleButtonText}
          </ToggleButton>
        ) : null}
      </div>
    );
  }
}

const ToggleButton = styled.span`
  color: ${props => props.theme.colors.primary.main};
  vertical-align: middle;
  font-size: 13px;
  line-height: 20px;
  padding: 0 5px;
  cursor: pointer;
`;
