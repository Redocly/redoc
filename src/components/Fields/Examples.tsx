import * as React from 'react';

import { FieldLabel, ExampleValue } from '../../common-elements/fields';
import { getSerializedValue, isArray } from '../../utils';

import { l } from '../../services/Labels';
import { FieldModel } from '../../services';
import styled from '../../styled-components';

export function Examples({ field }: { field: FieldModel }) {
  if (!field.examples) {
    return null;
  }

  return (
    <>
      <FieldLabel> {l('examples')}: </FieldLabel>
      {isArray(field.examples) ? (
        field.examples.map((example, idx) => {
          const value = getSerializedValue(field, example);
          const stringifyValue = field.in ? String(value) : JSON.stringify(value);
          return (
            <React.Fragment key={idx}>
              <ExampleValue>{stringifyValue}</ExampleValue>{' '}
            </React.Fragment>
          );
        })
      ) : (
        <ExamplesList>
          {Object.values(field.examples).map((example, idx) => (
            <li key={idx + example.value}>
              <ExampleValue>{getSerializedValue(field, example.value)}</ExampleValue> -{' '}
              {example.summary || example.description}
            </li>
          ))}
        </ExamplesList>
      )}
    </>
  );
}

const ExamplesList = styled.ul`
  margin-top: 1em;
  list-style-position: outside;
`;
