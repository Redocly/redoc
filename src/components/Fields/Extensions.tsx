import * as React from 'react';

import { ExtensionValue, FieldLabel } from '../../common-elements/fields';

import styled from '../../styled-components';

import { OptionsContext } from '../OptionsProvider';

import { StyledMarkdownBlock } from '../Markdown/styled.elements';

const Extension = styled(StyledMarkdownBlock)`
  margin: 2px 0;
`;

export interface ExtensionsProps {
  extensions: {
    [k: string]: any;
  };
}

export class Extensions extends React.PureComponent<ExtensionsProps> {
  render() {
    const exts = this.props.extensions;
    return (
      <OptionsContext.Consumer>
        {options => (
          <>
            {options.showExtensions &&
              Object.keys(exts).map(key => (
                <Extension key={key}>
                  <FieldLabel> {key.substring(2)}: </FieldLabel>{' '}
                  <ExtensionValue>
                    {typeof exts[key] === 'string' ? exts[key] : JSON.stringify(exts[key])}
                  </ExtensionValue>
                </Extension>
              ))}
          </>
        )}
      </OptionsContext.Consumer>
    );
  }
}
