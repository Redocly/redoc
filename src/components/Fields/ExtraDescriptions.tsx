import * as React from 'react';

import { ExtensionValue, FieldLabel } from '../../common-elements/fields';
import styled from '../../styled-components';

const Div = styled.div`
  padding-top: 2px;
  padding-bottom: 2px;

  &:last-child {
    padding-top: 2px;
    padding-bottom: 4px;
  }
`;

export interface ExtraDescriptionProps {
  extraDescriptions: Record<string, boolean>;
}

export class ExtraDescription extends React.PureComponent<ExtraDescriptionProps> {
  render() {
    const desc = this.props.extraDescriptions;

    return (
      <div>
        {Object.keys(desc).map(key => (
          <Div key={key}>
            <FieldLabel>{key}:</FieldLabel> <ExtensionValue>{desc[key]}</ExtensionValue>
          </Div>
        ))}
      </div>
    );
  }
}
