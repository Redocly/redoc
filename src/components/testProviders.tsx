import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import defaultTheme, { resolveTheme } from '../theme';

export default class TestThemeProvider extends React.Component {
  render() {
    return (
      <ThemeProvider theme={resolveTheme(defaultTheme)}>
        {React.Children.only(this.props.children as any)}
      </ThemeProvider>
    );
  }
}

export function withTheme(children) {
  return <TestThemeProvider>{children}</TestThemeProvider>;
}
