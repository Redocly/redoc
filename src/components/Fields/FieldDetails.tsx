import * as React from 'react';

import {
  PatternLabel,
  RecursiveLabel,
  TypeFormat,
  TypeName,
  TypePrefix,
  TypeTitle,
  ToggleButton,
  FieldLabel,
  ExampleValue,
} from '../../common-elements/fields';
import { serializeParameterValue } from '../../utils/openapi';
import { ExternalDocumentation } from '../ExternalDocumentation/ExternalDocumentation';
import { Markdown } from '../Markdown/Markdown';
import { EnumValues } from './EnumValues';
import { Extensions } from './Extensions';
import { FieldProps } from './Field';
import { ConstraintsView } from './FieldContstraints';
import { FieldDetail } from './FieldDetail';

import { Badge } from '../../common-elements/';

import { l } from '../../services/Labels';
import { OptionsContext } from '../OptionsProvider';
import { FieldModel } from '../../services/models/Field';
import styled from '../../styled-components';

const MAX_PATTERN_LENGTH = 45;

export class FieldDetails extends React.PureComponent<FieldProps, { patternShown: boolean }> {
  state = {
    patternShown: false,
  };

  static contextType = OptionsContext;

  togglePattern = () => {
    this.setState({
      patternShown: !this.state.patternShown,
    });
  };

  render() {
    const { showExamples, field, renderDiscriminatorSwitch } = this.props;
    const { patternShown } = this.state;
    const { enumSkipQuotes, hideSchemaTitles, hideSchemaPattern } = this.context;

    const { schema, description, example, deprecated, examples } = field;

    const rawDefault = !!enumSkipQuotes || field.in === 'header'; // having quotes around header field default values is confusing and inappropriate

    let renderedExamples: JSX.Element | null = null;

    if (showExamples && (example !== undefined || examples !== undefined)) {
      if (examples !== undefined) {
        renderedExamples = <Examples field={field} />;
      } else {
        const label = l('example') + ':';
        const raw = !!field.in;
        renderedExamples = <FieldDetail label={label} value={getSerializedValue(field, field.example)} raw={raw} />;
      }
    }

    return (
      <div>
        <div>
          <TypePrefix>{schema.typePrefix}</TypePrefix>
          <TypeName>{schema.displayType}</TypeName>
          {schema.displayFormat && (
            <TypeFormat>
              {' '}
              &lt;
              {schema.displayFormat}
              &gt;{' '}
            </TypeFormat>
          )}
          {schema.contentEncoding && (
            <TypeFormat>
              {' '}&lt;
              {schema.contentEncoding}
              &gt;{' '}
            </TypeFormat>
          )}
          {schema.contentMediaType && (
            <TypeFormat>
              {' '}&lt;
              {schema.contentMediaType}
              &gt;{' '}
            </TypeFormat>
          )}
          {schema.title && !hideSchemaTitles && <TypeTitle> ({schema.title}) </TypeTitle>}
          <ConstraintsView constraints={schema.constraints} />
          {schema.pattern && !hideSchemaPattern && (
            <>
              <PatternLabel>
                {patternShown || schema.pattern.length < MAX_PATTERN_LENGTH
                  ? schema.pattern
                  : `${schema.pattern.substr(0, MAX_PATTERN_LENGTH)}...`}
              </PatternLabel>
              {schema.pattern.length > MAX_PATTERN_LENGTH && (
                <ToggleButton onClick={this.togglePattern}>
                  {patternShown ? 'Hide pattern' : 'Show pattern'}
                </ToggleButton>
              )}
            </>
          )}
          {schema.isCircular && <RecursiveLabel> {l('recursive')} </RecursiveLabel>}
        </div>
        {deprecated && (
          <div>
            <Badge type="warning"> {l('deprecated')} </Badge>
          </div>
        )}
        <FieldDetail raw={rawDefault} label={l('default') + ':'} value={schema.default} />
        {!renderDiscriminatorSwitch && <EnumValues type={schema.type} values={schema.enum} />}{' '}
        {renderedExamples}
        {<Extensions extensions={{ ...field.extensions, ...schema.extensions }} />}
        <div>
          <Markdown compact={true} source={description} />
        </div>
        {schema.externalDocs && (
          <ExternalDocumentation externalDocs={schema.externalDocs} compact={true} />
        )}
        {(renderDiscriminatorSwitch && renderDiscriminatorSwitch(this.props)) || null}
        {field.const && (<FieldDetail label={l('const') + ':'} value={field.const} />) || null}
      </div>
    );
  }
}

function Examples({ field }: { field: FieldModel }) {
  if (!field.examples) {
    return null;
  }

  return (
    <>
      <FieldLabel> {l('examples')}: </FieldLabel>
      <ExamplesList>
        {Object.values(field.examples).map((example, idx) => {
          return (
            <li key={idx}>
              <ExampleValue>{getSerializedValue(field, example.value)}</ExampleValue> - {example.summary || example.description}
            </li>
          );
        })}
      </ExamplesList>
    </>
  );
}

function getSerializedValue(field: FieldModel, example: any) {
  if (field.in) {
    // decode for better readability in examples: see https://github.com/Redocly/redoc/issues/1138
    return decodeURIComponent(serializeParameterValue(field, example));
  } else {
    return example;
  }
}


const ExamplesList = styled.ul`
  margin-top: 1em;
  padding-left: 0;
  list-style-position: inside;
`;
