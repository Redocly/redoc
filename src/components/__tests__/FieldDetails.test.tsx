import * as React from 'react';
import { shallow } from 'enzyme';

import { FieldDetails } from '../Fields/FieldDetails';
import { SchemaModel } from '../../services/models/Schema';
import { withTheme } from '../testProviders';

jest.mock('../ExternalDocumentation/ExternalDocumentation', () => ({
  ExternalDocumentation: () => {
    return <div />;
  },
}));

describe('FieldDetailsComponent', () => {
  it('renders correctly', () => {
    const mockFieldProps = {
      showExamples: true,
      field: {
        schema: {
          type: 'array',
          default: [],
          typePrefix: 'test type prefix',
          displayType: 'array',
          title: 'test title',
          externalDocs: undefined,
          constraints: [''],
        } as SchemaModel,
        example: 'example',
        name: 'name',
        expanded: false,
        required: false,
        kind: '',
        deprecated: false,
        collapse: jest.fn(),
        toggle: jest.fn(),
        explode: false,
        expand: jest.fn(),
        description: 'test description',
      },
      renderDiscriminatorSwitch: jest.fn(),
    };

    const wrapper = shallow(withTheme(<FieldDetails {...mockFieldProps} />));

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('renders correctly when default value is object in request body', () => {
    const mockFieldProps = {
      showExamples: true,
      field: {
        schema: {
          type: 'object',
          default: { properties: {} },
          displayType: 'object',
          title: 'test title',
          externalDocs: undefined,
          constraints: [''],
        } as SchemaModel,
        example: 'example',
        name: 'name',
        expanded: false,
        required: false,
        kind: '',
        deprecated: false,
        collapse: jest.fn(),
        toggle: jest.fn(),
        explode: false,
        expand: jest.fn(),
        description: 'test description',
        in: undefined,
      },
      renderDiscriminatorSwitch: jest.fn(),
    };
    const wrapper = shallow(withTheme(<FieldDetails {...mockFieldProps} />));

    expect(wrapper.render()).toMatchSnapshot();
  });

  it('renders correctly when field items have string type and pattern', () => {
    const mockFieldProps = {
      showExamples: true,
      field: {
        schema: {
          type: 'array',
          displayType: 'Array of strings',
          title: 'test title',
          externalDocs: undefined,
          constraints: [''],
          items: {
            type: 'string',
            pattern: '^see regex[0-9]$',
            constraints: [''],
            externalDocs: undefined,
          },
        } as any as SchemaModel,
        example: 'example',
        name: 'name',
        expanded: false,
        required: false,
        kind: '',
        deprecated: false,
        collapse: jest.fn(),
        toggle: jest.fn(),
        explode: false,
        expand: jest.fn(),
        description: 'test description',
        in: undefined,
      },
      renderDiscriminatorSwitch: jest.fn(),
    };
    const wrapper = shallow(withTheme(<FieldDetails {...mockFieldProps} />));

    expect(wrapper.render()).toMatchSnapshot();
  });
});
