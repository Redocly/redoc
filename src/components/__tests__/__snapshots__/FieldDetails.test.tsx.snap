// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldDetailsComponent renders correctly 1`] = `
<div>
  <div>
    <span
      class="sc-kpDqfm sc-dAlyuH cGRfjn gHomYR"
    >
      test type prefix
    </span>
    <span
      class="sc-kpDqfm sc-jlZhew cGRfjn dYtiIA"
    >
      array
    </span>
    <span
      class="sc-kpDqfm sc-cwHptR cGRfjn gyVIPr"
    >
       (test title) 
    </span>
    <span>
       
      <span
        class="sc-kpDqfm sc-gFqAkR cGRfjn fYEICH"
      >
          
      </span>
    </span>
  </div>
  <div>
    <span
      class="sc-kpDqfm cGRfjn"
    >
       Default: 
    </span>
     
    <span
      class="sc-kpDqfm sc-eldPxv cGRfjn ehWiAn"
    >
      ""
    </span>
  </div>
   
  <div>
    <span
      class="sc-kpDqfm cGRfjn"
    >
       Example: 
    </span>
     
    <span
      class="sc-kpDqfm sc-eldPxv cGRfjn ehWiAn"
    >
      "example"
    </span>
  </div>
  <div>
    <div
      class="sc-lcIPJg sc-hknOHE gBHqkN jFBMaE"
    >
      <p>
        test description
      </p>
      

    </div>
  </div>
</div>
`;
